chalk-integration/src/lowering.rs

```diff
@@ -2082,84 +1736,46 @@ impl<'k> LowerGoal<Env<'k>> for Goal {
     }
 }
 
-trait LowerQuantifiedGoal {
-    fn lower_quantified(
-        &self,
-        env: &Env,
-        quantifier_kind: chalk_ir::QuantifierKind,
-        variable_kinds: &[VariableKind],
-    ) -> LowerResult<chalk_ir::Goal<ChalkIr>>;
-}
-
-impl LowerQuantifiedGoal for Goal {
-    fn lower_quantified(
-        &self,
-        env: &Env,
-        quantifier_kind: chalk_ir::QuantifierKind,
-        variable_kinds: &[VariableKind],
-    ) -> LowerResult<chalk_ir::Goal<ChalkIr>> {
-        let interner = env.interner();
-        if variable_kinds.is_empty() {
-            return self.lower(env);
-        }
-
-        let variable_kinds = variable_kinds.iter().map(|pk| pk.lower());
-        let subgoal = env.in_binders(variable_kinds, |env| self.lower(env))?;
-        Ok(chalk_ir::GoalData::Quantified(quantifier_kind, subgoal).intern(interner))
+fn lower_quantified(
+    goal: &Goal,
+    env: &Env,
+    quantifier_kind: chalk_ir::QuantifierKind,
+    variable_kinds: &[VariableKind],
+) -> LowerResult<chalk_ir::Goal<ChalkIr>> {
+    let interner = env.interner();
+    if variable_kinds.is_empty() {
+        return goal.lower(env);
     }
-}
 
-trait LowerWellKnownTrait {
-    fn lower(&self) -> rust_ir::WellKnownTrait;
+    let variable_kinds = variable_kinds.iter().map(lower_variable_kind);
+    let subgoal = env.in_binders(variable_kinds, |env| goal.lower(env))?;
+    Ok(chalk_ir::GoalData::Quantified(quantifier_kind, subgoal).intern(interner))
 }
 
-impl LowerWellKnownTrait for WellKnownTrait {
-    fn lower(&self) -> rust_ir::WellKnownTrait {
-        match self {
-            Self::Sized => rust_ir::WellKnownTrait::Sized,
-            Self::Copy => rust_ir::WellKnownTrait::Copy,
-            Self::Clone => rust_ir::WellKnownTrait::Clone,
-            Self::Drop => rust_ir::WellKnownTrait::Drop,
-            Self::FnOnce => rust_ir::WellKnownTrait::FnOnce,
-            Self::FnMut => rust_ir::WellKnownTrait::FnMut,
-            Self::Fn => rust_ir::WellKnownTrait::Fn,
-            Self::Unsize => rust_ir::WellKnownTrait::Unsize,
-        }
+fn apply_result<T>(result: LowerResult<Vec<T>>) -> Vec<LowerResult<T>> {
```

Comment:
This is a weird function. Wonder if this could be refactored away.

Reply:
I'll just inline it for now, couldn't really simplify it (surprisingly, there's no transpose for iterator/vec, and I couldn't find a nicer way)